Add `?introduce @user #help-solve`

DonaldKellett opened the pull request:
This command is only available to admins and mods.

- `?introduce @mention #help-solve`: DM's the mentioned user with guidelines on asking for help on solving Codewars Kata. Falls back to mentioning the user in the channel if DM is not available.

I have implemented most of the suggestions from the previous pull request, except that the messages are currently hardcoded in the source code instead of in separate text files. It was mentioned that the text files containing the messages should be in a separate directory, but which directory should we put it in? Once we've agreed on that, I can move the message contents out into a separate directory and fix any remaining issues.

DonaldKellett commented:
`?introduce @mention #off-topic` has been replaced with `?introduce @mention #anything` in response to the renaming of the affected channel.

kazk commented:
> It was mentioned that the text files containing the messages should be in a separate directory, but which directory should we put it in?

Messages can be under `text/` next to the command. I think that makes the most sense.

kazk commented on src/events/message/commands/help/introduce.ts:
```diff
@@ -0,0 +1,84 @@
+  }
+};
+
+const INTRO = `Welcome to the Codewars Discord server!
```
I don't think we need this. We have `#readme` and users are shown these when they join too.

kazk commented on src/events/message/commands/help/introduce.ts:
```diff
@@ -0,0 +1,84 @@
+\`\`\`
+5. The input, output from your solution and expected output`;
+
+const ANYTHING = `Welcome to #anything! You can discuss pretty much anything that is appropriate for all ages on this channel, whether related to programming or not. However, we ask you to respect a few rules:
```
Again, this isn't more than what everyone agreed on when joining the server and what's in `#readme`.

The only difference is

> - Nor is this a place to receive free help for completing your programming homework, project, assignment, etc. Similar measures apply as for advertising

What is the use case? I don't think we should be using this command when someone comments on that channel the first time. So, this must be used when they did something wrong. Then that should be a different command like `?warn` explaining what they did wrong.

DonaldKellett commented:
Thanks for the feedback! I suppose we just keep `?introduce @mention #help-solve` and perhaps add a `?warn` command in another pull request?

DonaldKellett commented:
I removed `?introduce @mention` and `?introduce @mention #anything` and moved the message contents for `?introduce @mention #help-solve` into a top-level directory `text/`. The reason I did not place `text/` alongside the source file for the command `introduce.ts` is because the compiled file `introduce.js` lives in a different directory from `introduce.ts` so I cannot just do `path.join(__dirname, 'text')` to access that directory (though it can be fixed by replacing `/src` with `/lib`). Is that fine by you?

kazk commented:
> The reason I did not place `text/` alongside the source file for the command `introduce.ts` is because ...

Oh I forgot about that, sorry. Yeah, that's fine.

kazk commented on src/events/message/commands/help/introduce.ts:
```diff
@@ -0,0 +1,47 @@
+    await dm.send(reply);
+  } catch (err) {
+    console.warn(`failed to DM ${user.tag}: ${err.message || "unknown error"}`);
+    await message.channel.send(`<@${mention.id}> ${reply}`);
```
I don't want the verbose text showing up in the channel. Just link to the file on GitHub.

```text
@user, I couldn't DM you. See <https://github.com/codewars/discord-bot/blob/text/help-solve.md> instead.
```

kazk commented on src/events/message/commands/help/introduce.ts:
```diff
@@ -0,0 +1,47 @@
+  const dmPath = path.join(path.join(process.cwd(), "text"), `${channel.name}.txt`);
+  let reply;
+  try {
+    reply = (await promises.readFile(dmPath)).toString();
```
Build a map of channel name to text once on load. It's not going to change.

kazk commented on text/help-solve.txt:
```diff
@@ -0,0 +1,17 @@
+Welcome to #help-solve! For an optimal experience, please include the following information **in #help-solve** (not to me!):
```
Change this to `.md`? It's easier to read when DM failed.

kazk commented on src/events/message/commands/help/introduce.ts:
```diff
@@ -0,0 +1,47 @@
+  // Input validation
+  if (args.length !== 2) return;
+  const mention = args[0];
+  if (typeof mention !== "object") return;
```
```suggestion
```
`CommandArg` is always an object.

kazk commented on src/events/message/commands/help/introduce.ts:
```diff
@@ -0,0 +1,47 @@
+  const user = message.client.users.cache.get(mention.id);
+  if (!user) return;
+  const channelMention = args[1];
+  if (typeof channelMention !== "object") return;
```
```suggestion
```

DonaldKellett commented:
Suggestions applied, please review

kazk commented on text/help-solve.md:
```diff
@@ -0,0 +1,18 @@
+3. What you have tried before asking for help. We'd love to put in the effort to help you, but only if you first help yourself.
+4. Your current solution in properly formatted code blocks. For example:
+
+\`\`\`javascript
+console.log("Hello World!");
+\`\`\`
```
```suggestion
\`\`\`javascript  
console.log("Hello World!");  
\`\`\`  
```

kazk commented on src/events/message/commands/help/introduce.ts:
```diff
@@ -0,0 +1,55 @@
+import { Message, CommandArg } from "../types";
+import { Role, TextChannel } from "discord.js";
+import { readFileSync } from "fs";
+import * as path from "path";
```
nit: I usually order imports like this: standard lib (Node.js modules), packages, then local. 

```suggestion
import { readFileSync } from "fs";
import * as path from "path";
import { Role, TextChannel } from "discord.js";
import { Message, CommandArg } from "../types";
```

kazk commented:
Thanks!
